crates/holochain/src/core/ribosome/host_fn.rs

```diff
@@ -1,32 +1,158 @@
-pub mod agent_info;
-pub mod call;
-pub mod call_remote;
-pub mod capability_claims;
-pub mod capability_grants;
-pub mod capability_info;
-pub mod create;
-pub mod create_link;
-pub mod debug;
-pub mod decrypt;
-pub mod delete;
-pub mod delete_link;
-pub mod emit_signal;
-pub mod encrypt;
-pub mod entry_type_properties;
-pub mod get;
-pub mod get_agent_activity;
-pub mod get_details;
-pub mod get_link_details;
-pub mod get_links;
-pub mod hash_entry;
-pub mod property;
-pub mod query;
-pub mod random_bytes;
-pub mod schedule;
-pub mod show_env;
-pub mod sign;
-pub mod sys_time;
-pub mod unreachable;
-pub mod update;
-pub mod verify_signature;
-pub mod zome_info;
+use super::{error::RibosomeError, CallContext, RibosomeT};
+use holochain_zome_types::zome_io::HostFnApiT;
+use std::sync::Arc;
+
+pub struct HostFnApi<Ribosome: RibosomeT> {
+    ribosome: Arc<Ribosome>,
+    call_context: Arc<CallContext>,
+}
+
+impl<Ribosome: RibosomeT> HostFnApi<Ribosome> {
+    pub fn new(ribosome: Arc<Ribosome>, call_context: Arc<CallContext>) -> Self {
+        Self {
+            ribosome,
+            call_context,
+        }
+    }
+}
+
+macro_rules! host_fn_api_impls {
+    ( $( fn $f:ident ( $input:ty ) -> $output:ty; )* ) => {
+        $(
+            pub(crate) mod $f;
+        )*
+
+        impl<Ribosome: RibosomeT> HostFnApiT for HostFnApi<Ribosome> {
+            type Error = RibosomeError;
+            $(
+                fn $f(&self, input: $input) -> Result<$output, Self::Error> {
+                    $f::$f(
+                        self.ribosome.clone(),
+                        self.call_context.clone(),
+                        input.into()
+                    ).map(|r| r.into_inner())
+                }
+            )*
+        }
+    };
+}
+
+// All host_fn_api_impls below rely on this import
+use holochain_zome_types as zt;
+
+host_fn_api_impls! {
+
+    // ------------------------------------------------------------------
+    // These definitions are copy-pasted from
+    // holochain_zome_types::zome_io
+    // TODO: is there a way to unhygienically import this code in both places?
```

Comment:
The next 100 lines of code now need to be copy-pasted from zome_io to here and kept in sync. This works because the two different macros take the same syntax but produce different code. The HostFnApiT trait will let us know if they ever *do* get out of sync, but anybody got ideas on how to DRY this up? `include!` doesn't work. Can we do this with a build.rs?

Reply:
I can't think of a way, it's not the worst thing to have it two places as it shouldn't change that frequently